Fix bug with correlate for tensorflow

danielenricocahall opened the pull request:
Addresses https://github.com/keras-team/keras/issues/21617 by ensuring the "full" and "same" case are properly handled. For simplicity, "same" just reuses "full" and truncates. To validate, a unit test has been added which uses the original data from the issue.

hertschuh commented on keras/src/backend/tensorflow/numpy.py:
```diff
@@ -3130,30 +3130,47 @@ def correlate(x1, x2, mode="valid"):
+        # b: filter [M] -> [M,1,1]
+        return (
+            tf.reshape(a, (1, tf.shape(a)[0], 1)),
+            tf.reshape(b, (tf.shape(b)[0], 1, 1)),
```
Please use `shape_op(a)` / `shape_op(b)` instead of `tf.shape`.

The reason is that `shape_op` returns ints when the shape is static, whereas `tf.shape` always returns a tensor, which causes all subsequent shapes to be dynamic.

hertschuh commented on keras/src/backend/tensorflow/numpy.py:
```diff
@@ -3130,30 +3130,47 @@ def correlate(x1, x2, mode="valid"):
+        x1 = tf.pad(x1, [[pad, pad]])  # pad input with zeros
+        x1, x2 = _pack(x1, x2)
+        out = tf.nn.conv1d(x1, x2, stride=1, padding="VALID")
+        return tf.squeeze(out)
```
I realize you didn't code that initially, but can we be specific about which dimensions to squeeze? My concern is if the length of the output is 1, we'll accidentally squeeze a dimension that we shouldn't.

Additionally, if we don't do that, it means that the rank of the output cannot always be statically determined if the outputs size is dynamic.

hertschuh commented on keras/src/backend/tensorflow/numpy.py:
```diff
@@ -3130,30 +3130,47 @@ def correlate(x1, x2, mode="valid"):
+        return tf.slice(full_corr, [start], [out_len])
+    elif mode == "valid":
+        x1, x2 = _pack(x1, x2)
+        return tf.squeeze(tf.nn.conv1d(x1, x2, stride=1, padding="VALID"))
```
I realize you didn't code that initially, but can we be specific about which dimensions to squeeze? My concern is if the length of the output is 1, we'll accidentally squeeze a dimension that we shouldn't.

Additionally, if we don't do that, it means that the rank of the output cannot always be statically determined if the outputs size is dynamic.

hertschuh commented on keras/src/ops/numpy_test.py:
```diff
@@ -5133,6 +5133,26 @@ def test_correlate_different_size(self):
             knp.Correlate(mode="full")(x, y), np.correlate(x, y, mode="full")
         )
 
+    def test_correlate_bug(self):
```
Since these cover a very similar scenario as `test_correlate_different_size`, the test above, let's replace `test_correlate_different_size` with this one. So remove `test_correlate_different_size` and rename this one `test_correlate_different_size`.

hertschuh reviewed (commented):
@danielenricocahall 

Thank you for fixing this! This was a tricky one.

danielenricocahall commented on keras/src/backend/tensorflow/numpy.py:
```diff
@@ -3130,30 +3130,47 @@ def correlate(x1, x2, mode="valid"):
+    def _full_corr(x1, x2):
+        """Compute 'full' correlation result (length = n + m - 1)."""
+        m = tf.shape(x2)[0]
+        pad = tf.maximum(m - 1, 0)
```
Thank you! done

danielenricocahall commented:
Hey @hertschuh thank you for the feedback! I think I've addressed all comments - to confirm, for `squeeze`, we wanted to be explicit about only dropping the first and last (batch and channel) axes to ensure we don't squeeze the correlation result if it winds up being of size 1?

hertschuh commented:
> Hey @hertschuh thank you for the feedback! I think I've addressed all comments - to confirm, for `squeeze`, we wanted to be explicit about only dropping the first and last (batch and channel) axes to ensure we don't squeeze the correlation result if it winds up being of size 1?

Yes, exactly!

hertschuh commented on keras/src/backend/tensorflow/numpy.py:
```diff
@@ -3130,30 +3130,57 @@ def correlate(x1, x2, mode="valid"):
+        out_len = (
+            max(n, m)
+            if isinstance(n, int) and isinstance(m, int)
+            else tf.maximum(n, m)
```
`tf.maximum([n, m])`

danielenricocahall replied:
ah yes it's `max([n, m])` since it's not the builtin `max`, it's the one from the module. Fixed!
